build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt

```diff
@@ -402,20 +391,29 @@ open class IncrementalJvmCache(
                 storage.remove(key)
             }
 
-            for (const in oldMap.keys + newMap.keys) {
-                //Constant can be declared via companion object or via const field declaration
-                changesCollector.collectMemberIfValueWasChanged(
-                    kotlinClassInfo.scopeFqName(companion = true),
-                    const,
-                    oldMap[const],
-                    newMap[const]
-                )
-                changesCollector.collectMemberIfValueWasChanged(
-                    kotlinClassInfo.scopeFqName(companion = false),
-                    const,
-                    oldMap[const],
-                    newMap[const]
-                )
+            val allConstants = oldMap.keys + newMap.keys
+            if (allConstants.isEmpty()) return
+
+            // If a constant is defined in a companion object, it will be found in the constantsMap of the containing class, not the
+            // companion object's class, so we will need to correct its scope.
+            // (See https://youtrack.jetbrains.com/issue/KT-44741#focus=Comments-27-5659564.0-0 for more details.)
+            // Note: This only applies to a *constant* defined in a *companion object* (it's not an issue for inline functions, or top-level
+            // constants, or constants in non-companion objects).
+            val companionObjectClassId = if (kotlinClassInfo.classKind == KotlinClassHeader.Kind.CLASS) {
```

Comment:
Can you pls add a test case in `jps-plugin/testData/incremental/pureKotlin/companionConstantChanged/A.kt` which was added in the initial fix?

Reply:
I'm not familiar with that code base so I just added a minimal change: Updating the existing test data so that the companion object has a custom name. If you think it deserves a separate test, I'll probably ask Nataliya to help.

Reply:
I see you've already added it in e6be2cc1dee55d9eb799f681b3d086960c845215, so it LGTM.